Команда 13 реализовала приложение Filmorate

MaximusDenisov-YP opened the pull request.

MaximusDenisov-YP committed:
Реализовал функционал "Отзывы"

- Добавлена модель MVC для взаимодействия с функционалом.
- Изменены названия переменных, классов и файлов в соответствии с договорённостями в части нейминга по БД.
- Like переименован в FilmLike, интерфейс и реализация репозитория помещены в одну директорию, история с Review аналогичная.

MaximusDenisov-YP committed:
Реализовал функционал "Популярные фильмы по жанру и годам"

- Изменена ручка в /films/popular, добавлена возможность сортировки по genreId и year.
- Добавлен метод получения популярных фильмов по жанру и году выпуска из БД.
- Расширен интерфейс FilmStorage, добавлен шаг который указан выше.

MaximusDenisov-YP committed:
Удалил забытую, лишнюю логику из логики ReviewLikeDbStorage.java

- Добавил актуальное изображение диаграммы БД
- Подкорректировал README.md

MaximusDenisov-YP committed:
Merge pull request #5 from MaximusDenisov-YP/add-reviews

Реализовал функционал "Отзывы"

MaximusDenisov-YP committed:
Причесал реализацию по функциональности "Популярных фильмов по жанру и году"

- Изменил вызов в контроллере, так как оба метода вызываются по одной ручке, а в контроллере не должно быть бизнес логики, вынес эту логику на уровень Service

Commit:
реализовал функциональность «Рекомендации»

Commit:
реализовал функциональность «Рекомендации»

MaximusDenisov-YP committed:
Merge pull request #8 from MaximusDenisov-YP/add-most-populars

Реализовал функционал "Популярные фильмы по жанру и годам"

MaximusDenisov-YP committed:
Merge remote-tracking branch 'origin/add-recommendations' into add-recommendations

MaximusDenisov-YP committed:
Merge pull request #9 from MaximusDenisov-YP/add-recommendations

реализовал функциональность «Рекомендации»

SergeyLeshchev committed:
Реализовать функцию удаления фильма и пользователя по id

SergeyLeshchev committed:
Исправление ошибки Checkstyle

Commit:
1Описание задачи
 В информацию о фильмах должно быть добавлено имя режиссёра. После этого должна появиться следующая функциональность.

 Вывод всех фильмов режиссёра, отсортированных по количеству лайков.
 Вывод всех фильмов режиссёра, отсортированных по годам.
 API
 GET /films/director/{directorId}?sortBy=[year,likes]

 Возвращает список фильмов режиссера отсортированных по количеству лайков или году выпуска.

 POST /films

 {
 	"name": "New film",
 	"releaseDate": "1999-04-30",
 	"description": "New film about friends",
 	"duration": 120,
 	"mpa": { "id": 3},
 	"genres": [{ "id": 1}],
 	"director": [{ "id": 1}]
 }
 GET /directors - Список всех режиссёров

 GET /directors/{id}- Получение режиссёра по id

 POST /directors - Создание режиссёра

 PUT /directors - Изменение режиссёра

 {
 "id": 1,
 "name": "New director"
 }
 DELETE /directors/{id} - Удаление режиссёр

Commit:
Merge branch 'develop_rpavlov_directors' of https://github.com/MaximusDenisov-YP/java-filmorate into add-director

# Conflicts:
#	src/main/resources/schema.sql

Commit:
Решены конфликты мерджа.

MaximusDenisov-YP committed:
Исправлены проверки по Checkstyle

MaximusDenisov-YP committed:
Merge pull request #10 from MaximusDenisov-YP/add-remove-endpoint

Реализовать функцию удаления фильма и пользователя по id

Commit:
Merge branch 'develop' of https://github.com/MaximusDenisov-YP/java-filmorate into add-director

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java

MaximusDenisov-YP committed:
Merge pull request #11 from MaximusDenisov-YP/add-director

Add director

Commit:
реализовал функциональность Общие фильмы

Commit:
тесты

Commit:
тесты checkstyle

MaximusDenisov-YP committed:
Merge pull request #12 from MaximusDenisov-YP/add-common-films

реализовал функциональность Общие фильмы

SergeyLeshchev committed:
Добавить функционал Лента событий

MaximusDenisov-YP committed:
Merge pull request #14 from MaximusDenisov-YP/add-feed

Добавить функционал Лента событий

Commit:
Функциональность «Поиск». 3SP
💡 Ветка для реализации задачи должна называться add-search.

Описание задачи
Нужно реализовать поиск по названию фильмов и по режиссёру (если вы решили добавить информацию о режиссёрах).

Алгоритм должен уметь искать по подстроке. Например, вы вводите «крад», а в поиске возвращаются следующие фильмы: «Крадущийся тигр, затаившийся дракон», «Крадущийся в ночи» и другие.

API
GET /fimls/search

Возвращает список фильмов, отсортированных по популярности.

Параметры строки запроса

query — текст для поиска

by — может принимать значения director (поиск по режиссёру), title (поиск по названию), либо оба значения через запятую при поиске одновременно и по режиссеру и по названию.

Пример

GET /films/search?query=крад&by=director,title

Commit:
Функциональность «Поиск». 3SP
💡 Ветка для реализации задачи должна называться add-search.

Описание задачи
Нужно реализовать поиск по названию фильмов и по режиссёру (если вы решили добавить информацию о режиссёрах).

Алгоритм должен уметь искать по подстроке. Например, вы вводите «крад», а в поиске возвращаются следующие фильмы: «Крадущийся тигр, затаившийся дракон», «Крадущийся в ночи» и другие.

API
GET /fimls/search

Возвращает список фильмов, отсортированных по популярности.

Параметры строки запроса

query — текст для поиска

by — может принимать значения director (поиск по режиссёру), title (поиск по названию), либо оба значения через запятую при поиске одновременно и по режиссеру и по названию.

Пример

GET /films/search?query=крад&by=director,title

Commit:
Функциональность «Поиск». 3SP
💡 Ветка для реализации задачи должна называться add-search.

Описание задачи
Нужно реализовать поиск по названию фильмов и по режиссёру (если вы решили добавить информацию о режиссёрах).

Алгоритм должен уметь искать по подстроке. Например, вы вводите «крад», а в поиске возвращаются следующие фильмы: «Крадущийся тигр, затаившийся дракон», «Крадущийся в ночи» и другие.

API
GET /fimls/search

Возвращает список фильмов, отсортированных по популярности.

Параметры строки запроса

query — текст для поиска

by — может принимать значения director (поиск по режиссёру), title (поиск по названию), либо оба значения через запятую при поиске одновременно и по режиссеру и по названию.

Пример

GET /films/search?query=крад&by=director,title

RuslanPavlov9 committed:
Merge pull request #18 from MaximusDenisov-YP/add-search

Функциональность «Поиск». 3SP

MaximusDenisov-YP committed:
Исправил функциональность в соответствии с общими тестами по проекту

MaximusDenisov-YP committed:
Merge pull request #19 from MaximusDenisov-YP/add-reviews

Исправил функциональность в соответствии с общими тестами по проекту

Commit:
Поправил созависимые ошибки. Добавил жанры в список и отдаю 404 если фильм по режиссеру не найден.

MaximusDenisov-YP committed:
Merge pull request #20 from MaximusDenisov-YP/add-search

Поправил созависимые ошибки. Добавил жанры в список и отдаю 404 если …

MaximusDenisov-YP committed:
Исправил функциональность рекомендаций в соответствии с общими тестами

SergeyLeshchev committed:
Исправить ошибки в прохождении тестов постман

1. Добавить проверку существования пользователя в getEvents
2. Изменить поле Event с id на eventId
3. Убрать лишний код в удалении пользователя

MaximusDenisov-YP committed:
Merge pull request #21 from MaximusDenisov-YP/add-recommendations

Исправил функциональность рекомендаций в соответствии с общими тестами

SergeyLeshchev committed:
Merge branch 'develop' into add-feed

MaximusDenisov-YP committed:
Исправлена функциональность логирования в соответствие с тест-кейсами

MaximusDenisov-YP committed:
Исправлены замечания по CheckStyle

MaximusDenisov-YP committed:
Merge pull request #22 from MaximusDenisov-YP/add-feed

Add feed

avfyodorov commented on README.md:
```diff
@@ -1,8 +1,9 @@
 # java-filmorate
-Template repository for Filmorate project.
+<H3>Проект Java-Filmorate - Команда №13</H3>
```
добавьте, пожалуйста, в readme файл кто какую функциональность реализовал.

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/model/Review.java:
```diff
@@ -0,0 +1,20 @@
+    @NotNull
+    Long filmId;
+    int useful;
+}
```
Для соблюдения инкапсуляции все поля лучше сделать приватными.

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/exception/GlobalExceptionHandler.java:
```diff
@@ -57,6 +57,18 @@ public ResponseEntity<ErrorResponse> handleNotFoundException(NotFoundException e
         );
     }
 
+    @ExceptionHandler(NotFoundExceptionDirectors.class)
```
Кстати, лучше создать один класс исключение NotFoundException и в описании ошибки уже писать с чем связано исключение с user, director и тд
 в больших проектах может быть сотни  сущностей, зачем на каждую из них создавать отдельное исключение. Лучше сделать по одному исключению на каждый тип ошибки, а не сущности.

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/storage/director/DirectorDbStorage.java:
```diff
@@ -0,0 +1,148 @@
+
+        List<Film> films = jdbcTemplate.query(sql, (rs, rowNum) -> mapRowToFilm(rs), id, sortBy, sortBy);
+
+        films.forEach(film -> {
```
```java
   films.forEach(film -> {
            List<Director> directors = getDirectorsByFilmId(film.getId());
.......
            List<Genre> genres = getGenresByFilmId(film.getId());
.......
```
В этом месте, в цикле,  идут запросы к базе данных. Было бы хорошо избавиться от этих многочисленных запросов, которые сильно нагружают сервер. Хорошо бы переделать данный метод, чтобы уменьшить число обращений к БД .

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -0,0 +1,157 @@
+    public void setLike(long reviewId, long userId) {
+        if (hasUserLike(reviewId, userId)) {
+            reviewLikeStorage.removeReviewLike(reviewId, userId);
+//            recalcLikesForReview(reviewId);
```
Здесь и дальше.
Если код больше не нужен, его лучше удалить.

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -27,6 +28,36 @@ public class FilmDbStorage implements FilmStorage {
 
     private final JdbcTemplate jdbcTemplate;
 
+    private static final String GET_COMMON_FILMS_SQL =
```
Обратил внимание на запросы - они заметно отличаются в лучшую сторону, Здорово!

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/controller/DirectorController.java:
```diff
@@ -0,0 +1,50 @@
+
+@RestController
+@RequestMapping("/directors")
+@Validated
```
`@Validated` не имеет смысла, так как отсутствуют аннотации валидации(ограничения) для параметров методов)

avfyodorov reviewed (commented):
Добрый вечер! 

У вас очень хорошая работа. Вы быстро сумели сделать грамотную реализацию задач, с разбивкой по слоям и разделением функционала. А, с учётом того, что команда у вас несыгранная,  результат можно считать просто прекрасным.   Вам удалось избежать типичных ошибок, возникающих на этом этапе.
Отметил качество SQL запросов )

От меня только несколько мелких уточнений.

MaximusDenisov-YP committed:
Исправил замечания ревьюера

- Удалил @Validated из DirectorControleller, так как в нём больше нет проверочных аннотаций, которым он необходим.
- Удалил лишние ErrorResponseDirectors и NotFoundExceptionDirectors по факту отсутствия в их необходимости.
- Убрал хендлер эксепшиона выше.
- Добавлены забытые модификаторы доступа для свойств класса Review.
- Почистил файл ReviewService от закомментированного кода.
- Исправлен O(n) запросы в БД для получения фильмов с режиссёрами, теперь всё получаем всё в рамках одного запроса.

MaximusDenisov-YP committed:
Merge pull request #24 from MaximusDenisov-YP/add-reviews

Исправил замечания ревьюера

avfyodorov reviewed (commented):
Добрый день! 

Замечаний нет. 
Групповой проект принят. 
 Поздравляю с удачным опытом групповой разработки, успехов!
